File: src/api/control.rs
use axum::{
    extract::{Path, State},
    Json,
};
use serde::Deserialize;
use serde_json::Value;
use uuid::Uuid;

use crate::{
    auth::middleware::AuthUser,
    error::{AppError, Result},
    models::{
        audit::AuditLog,
        claim::ControlClaim,
        device::Device,
    },
    ws::commands::DeviceCommand,
    AppState,
};

async fn get_device_for_user(device_id: Uuid, org_id: Uuid, state: &AppState) -> Result<Device> {
    let device = Device::find_by_id(device_id, &state.db)
        .await?
        .ok_or(AppError::NotFound)?;
    if device.org_id != Some(org_id) {
        return Err(AppError::NotFound);
    }
    Ok(device)
}

pub async fn claim_control(
    State(state): State<AppState>,
    AuthUser(claims): AuthUser,
    Path(device_id): Path<Uuid>,
) -> Result<Json<serde_json::Value>> {
    get_device_for_user(device_id, claims.org_id, &state).await?;

    let claim = ControlClaim::claim(device_id, claims.sub, state.config.control.claim_ttl, &state.db).await?;

    let _ = AuditLog::log_user_command(&state.db, claims.sub, device_id, "claim.acquire", None).await;

    Ok(Json(serde_json::json!({
        "claimed": true,
        "expires_at": claim.expires_at,
    })))
}

pub async fn release_control(
    State(state): State<AppState>,
    AuthUser(claims): AuthUser,
    Path(device_id): Path<Uuid>,
) -> Result<Json<serde_json::Value>> {
    get_device_for_user(device_id, claims.org_id, &state).await?;

    // Admins can force-release any claim
    if claims.role == "admin" {
        ControlClaim::force_release(device_id, &state.db).await?;
    } else {
        ControlClaim::release(device_id, claims.sub, &state.db).await?;
    }

    let _ = AuditLog::log_user_command(&state.db, claims.sub, device_id, "claim.release", None).await;

    Ok(Json(serde_json::json!({ "released": true })))
}

pub async fn send_command(
    State(state): State<AppState>,
    AuthUser(claims): AuthUser,
    Path(device_id): Path<Uuid>,
    Json(body): Json<Value>,
) -> Result<Json<serde_json::Value>> {
    get_device_for_user(device_id, claims.org_id, &state).await?;

    // Validate claim (admins bypass)
    if claims.role != "admin" {
        ControlClaim::validate_active(device_id, claims.sub, &state.db).await?;
    }

    // Parse and validate the command
    let command: DeviceCommand = serde_json::from_value(body.clone())
        .map_err(|e| AppError::InvalidCommand(e.to_string()))?;
    command.validate()?;

    // Publish to Redis
    let channel = format!("commands:{device_id}");
    let wire_json = command.to_wire_json().to_string();

    let mut redis_conn = state.redis.clone();
    redis::cmd("PUBLISH")
        .arg(&channel)
        .arg(&wire_json)
        .query_async::<()>(&mut redis_conn)
        .await?;

    // Refresh claim expiry on successful command
    if claims.role != "admin" {
        let _ = ControlClaim::refresh_expiry(device_id, claims.sub, state.config.control.claim_ttl, &state.db).await;
    }

    // Audit log
    let _ = AuditLog::log_user_command(
        &state.db,
        claims.sub,
        device_id,
        "command.send",
        Some(body),
    )
    .await;

    Ok(Json(serde_json::json!({ "status": "sent" })))
}

File: src/api/control.ts
import { Hono } from 'hono'
import type { AppEnv } from '../types.js'
import { AppError } from '../error.js'
import { authMiddleware } from '../auth/middleware.js'
import { validateCommand, toWireJson } from '../ws/commands.js'

const control = new Hono<AppEnv>()

control.use(authMiddleware)

async function getDeviceForUser(deviceId: string, orgId: string, state: any) {
  const [device] = await state.db`SELECT * FROM devices WHERE id = ${deviceId}`
  if (!device || device.org_id !== orgId) throw AppError.notFound()
  return device
}

control.post('/:id/control/claim', async (c) => {
  const { state, user } = c.var
  const deviceId = c.req.param('id')
  await getDeviceForUser(deviceId, user.org_id, state)

  const claimTtl = state.config.control.claim_ttl
  const expiresAt = new Date(Date.now() + claimTtl * 1000)

  // Check existing claim
  const [existing] = await state.db`
    SELECT * FROM control_claims WHERE device_id = ${deviceId}
  `

  if (existing && new Date(existing.expires_at) > new Date() && existing.user_id !== user.sub) {
    // Active claim by another user
    const [claimUser] = await state.db`SELECT display_name FROM users WHERE id = ${existing.user_id}`
    throw AppError.deviceClaimed(existing.user_id, claimUser?.display_name || '', new Date(existing.expires_at))
  }

  let claim
  if (existing && existing.user_id === user.sub) {
    // Same user — refresh
    ;[claim] = await state.db`
      UPDATE control_claims SET expires_at = ${expiresAt} WHERE id = ${existing.id}
      RETURNING *
    `
  } else {
    // New claim (delete any stale)
    await state.db`DELETE FROM control_claims WHERE device_id = ${deviceId}`
    ;[claim] = await state.db`
      INSERT INTO control_claims (device_id, user_id, expires_at)
      VALUES (${deviceId}, ${user.sub}, ${expiresAt})
      RETURNING *
    `
  }

  await state.db`
    INSERT INTO audit_log (actor_type, actor_id, action, target_type, target_id)
    VALUES ('user', ${user.sub}, 'claim.acquire', 'device', ${deviceId})
  `.catch(() => {})

  return c.json({ claimed: true, expires_at: claim.expires_at })
})

control.post('/:id/control/release', async (c) => {
  const { state, user } = c.var
  const deviceId = c.req.param('id')
  await getDeviceForUser(deviceId, user.org_id, state)

  if (user.role === 'admin') {
    await state.db`DELETE FROM control_claims WHERE device_id = ${deviceId}`
  } else {
    await state.db`DELETE FROM control_claims WHERE device_id = ${deviceId} AND user_id = ${user.sub}`
  }

  await state.db`
    INSERT INTO audit_log (actor_type, actor_id, action, target_type, target_id)
    VALUES ('user', ${user.sub}, 'claim.release', 'device', ${deviceId})
  `.catch(() => {})

  return c.json({ released: true })
})

control.post('/:id/control/command', async (c) => {
  const { state, user } = c.var
  const deviceId = c.req.param('id')
  await getDeviceForUser(deviceId, user.org_id, state)

  // Validate claim (admins bypass)
  if (user.role !== 'admin') {
    const [claim] = await state.db`
      SELECT * FROM control_claims WHERE device_id = ${deviceId}
    `
    if (!claim || claim.user_id !== user.sub || new Date(claim.expires_at) <= new Date()) {
      throw AppError.noControlClaim()
    }
  }

  const body = await c.req.json()
  validateCommand(body)

  // Publish to Redis
  const channel = `commands:${deviceId}`
  const wireJson = JSON.stringify(toWireJson(body))
  await state.redis.publish(channel, wireJson)

  // Refresh claim TTL
  if (user.role !== 'admin') {
    const newExpiry = new Date(Date.now() + state.config.control.claim_ttl * 1000)
    await state.db`
      UPDATE control_claims SET expires_at = ${newExpiry}
      WHERE device_id = ${deviceId} AND user_id = ${user.sub}
    `.catch(() => {})
  }

  await state.db`
    INSERT INTO audit_log (actor_type, actor_id, action, target_type, target_id, details)
    VALUES ('user', ${user.sub}, 'command.send', 'device', ${deviceId}, ${JSON.stringify(body)})
  `.catch(() => {})

  return c.json({ status: 'sent' })
})

export default control
